keras_nlp/layers/transformer_decoder.py

```diff
@@ -0,0 +1,229 @@
+        x = self._intermediate_dense(input)
+        x = self._intermediate_dropout(x)
+        x = self._output_dense(x)
+        return self._output_dropout(x)
+
+    def call(
+        self,
+        decoder_sequence,
+        encoder_sequence,
+        decoder_padding_mask=None,
+        decoder_additional_mask=None,
+        encoder_padding_mask=None,
+        encoder_additional_mask=None,
+        use_causal_mask=False,
+    ):
+        """Forward pass of the TransformerDecoder.
+
+        Args:
+            decoder_sequence: a Tensor. The decoder input sequence.
+            encoder_sequence: a Tensor. The decoder input sequence.
+            decoder_padding_mask: a boolean Tensor, the padding mask of decoder
+                sequence, must of shape [batch_size, decoder_sequence_length].
+            decoder_additional_mask: a boolean Tensor. Customized decoder
+                sequence mask, must of shape
+                [batch_size, decoder_sequence_length, decoder_sequence_length].
+            encoder_padding_mask: a boolean Tensor, the padding mask of encoder
+                sequence, must of shape [batch_size, encoder_sequence_length].
+            encoder_additional_mask: a boolean Tensor. Customized encoder
+                sequence mask, must of shape
+                [batch_size, encoder_sequence_length, encoder_sequence_length].
+            use_causal_mask: bool, defaults to False. If true, causal mask
+                (masking out future input) is applied on the decoder sequence.
+
+        Returns:
+            A Tensor of the same shape as the `decoder_sequence`.
+        """
+        if not self._built:
+            self._build(decoder_sequence.shape)
+        encoder_mask = merge_padding_and_additional_mask(
+            encoder_sequence, encoder_padding_mask, encoder_additional_mask
+        )
+        decoder_mask = merge_padding_and_additional_mask(
+            decoder_sequence, decoder_padding_mask, decoder_additional_mask
+        )
+        if use_causal_mask:
+            causal_mask = tf.cast(
+                compute_causal_mask(decoder_sequence),
+                dtype=tf.int32,
+            )
+            if decoder_mask is None:
+                decoder_mask = causal_mask
+            else:
+                decoder_mask = tf.minimum(decoder_mask, causal_mask)
+
+        # Decoder input self-attention.
+        self_attended = self._self_attention(decoder_sequence, decoder_mask)
+        self_attended = self._add_and_norm(
+            self_attended, decoder_sequence, self._decoder_attention_layernorm
+        )
+        # Encoder-decoder attention.
+        encoder_decoder_attended = self._encoder_decoder_attention(
+            encoder_sequence, self_attended, encoder_mask
+        )
+        encoder_decoder_attended = self._add_and_norm(
+            encoder_decoder_attended,
+            self_attended,
+            self._enc_dec_attention_layernorm,
+        )
+        # Feedforward and add&norm.
```

Comment:
nit: # Feedforward block.

(you don't mention add and norm for the other block comments above)

Reply:
Done